Decouple BitmapImage sources and main functions into separate classes

shallawa opened the pull request:
#### 8b78e07f7be5805e58bc1858db1ee8a6e6a7a15d
<pre>
Decouple BitmapImage sources and main functions into separate classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=265363">https://bugs.webkit.org/show_bug.cgi?id=265363</a>
<a href="https://rdar.apple.com/118819338">rdar://118819338</a>

Reviewed by Cameron McCormack.

The BitmapImage and ImageSource manage the operations and the queries of the
bitmap image. The current interface and the code structure of these two classes
and the relationship between them make it difficult and risky to hack.

This purpose of this refactor is to provide a clearer interface and to build a
a better class hierarchy and association. This will allow distributing the
operations among better encapsulated classes.

Here is the new class hierarchy and association:

  Image: The super class for all kinds of images in WebKit.

  ImageAdapter: It converts an Image to a platform specific format.

  ImageDecoder: A super class for the system and WebKit image decoders

  ImageSource: A super class BitmapImage source. It can be from a CachedImage or
  from a NativeImage.

  BitmapImageSource: Receives the encoded data of an image and provides all the
  frames decoded NativeImage and metadata. It holds and coordinates the interactions
  amog BitmapImage, ImageFrameAnimator and ImageFrameWorkQueue.

  NativeImageSource: It wraps an already created NativeImage in a BitmapImage.

  ImageFrameAnimator: It advances the current frame index of an animated image when.
  its timer fires. The timer is set with the duration of the frame. It requests
  decoding the NativeImage asynchronously if this is allowed.

  ImageFrameWorkQueue: It decodes the NativeImage of an ImageFrame asynchronously
  on a separate WorkQueue. Once the NativeImage is ready, it asks BitmapImageSource
  to cache it on the main thread. BitmapImageSource will notifies the observer
  that a new frame is ready for rendering.

  BitmapImage: This is a sub-class of Image which mainly handles the layout and
  the drawing of a bitmap image to a GraphicsContext.

An important change of this PR is the DecodingMode and AllowImageSubsampling are
calculated in advance before calling BitmapImage::draw(). So Settings are no
longer passed or updated by BitmapImage before calling its draw() method.

* LayoutTests/fast/images/async-image-background-image-repeated.html:
Make the frame decoding duration time longer than the animation duration time so
the decoding always finishes after the animation timer fires.

* LayoutTests/fast/images/async-image-composited-show.html:
Make this test do the decoding asynchronously. The debug settings were missing.

* LayoutTests/fast/images/async-image-intersect-different-size-for-drawing-expected.txt:
If the decodingMode is Synchronous, no sizeForDrawing should be set.

* LayoutTests/fast/images/decode-render-animated-image-expected.html:
* LayoutTests/fast/images/decode-render-animated-image.html:
Make the test relies solely on the `decode()` method instead of the mix between
the debug event &quot;webkitImageFrameReady&quot; and the `decode()` method.

* LayoutTests/platform/ios-wk2/fast/images/async-image-intersect-different-size-for-drawing-expected.txt:
If the decodingMode is Synchronous, no sizeForDrawing should be set.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::decode):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::removeAllClientsWaitingForAsyncDecoding):
(WebCore::CachedImage::CachedImageObserver::canDestroyDecodedData const):
(WebCore::CachedImage::canDestroyDecodedData const):
(WebCore::CachedImage::CachedImageObserver::canDestroyDecodedData): Deleted.
(WebCore::CachedImage::canDestroyDecodedData): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::canDestroyDecodedData const):
(WebCore::CachedImageClient::canDestroyDecodedData): Deleted.
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::createDragImageFromImage):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::create):
(WebCore::BitmapImage::BitmapImage):
(WebCore::BitmapImage::dataChanged):
(WebCore::BitmapImage::destroyDecodedData):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::drawPattern):
(WebCore::BitmapImage::dump const):
(WebCore::BitmapImage::~BitmapImage): Deleted.
(WebCore::BitmapImage::destroyDecodedDataIfNecessary): Deleted.
(WebCore::BitmapImage::setCurrentFrameDecodingStatusIfNecessary): Deleted.
(WebCore::BitmapImage::nativeImageAtIndexCacheIfNeeded): Deleted.
(WebCore::BitmapImage::nativeImage): Deleted.
(WebCore::BitmapImage::nativeImageForCurrentFrame): Deleted.
(WebCore::BitmapImage::preTransformedNativeImageForCurrentFrame): Deleted.
(WebCore::BitmapImage::notSolidColor): Deleted.
(WebCore::drawNativeImage): Deleted.
(WebCore::BitmapImage::shouldAnimate const): Deleted.
(WebCore::BitmapImage::canAnimate const): Deleted.
(WebCore::BitmapImage::canUseAsyncDecodingForLargeImages const): Deleted.
(WebCore::BitmapImage::shouldUseAsyncDecodingForAnimatedImages const): Deleted.
(WebCore::BitmapImage::clearTimer): Deleted.
(WebCore::BitmapImage::startTimer): Deleted.
(WebCore::BitmapImage::subsamplingLevelForScaleFactor): Deleted.
(WebCore::BitmapImage::canDestroyDecodedData): Deleted.
(WebCore::BitmapImage::internalStartAnimation): Deleted.
(WebCore::BitmapImage::advanceAnimation): Deleted.
(WebCore::BitmapImage::internalAdvanceAnimation): Deleted.
(WebCore::BitmapImage::isAnimating const): Deleted.
(WebCore::BitmapImage::stopAnimation): Deleted.
(WebCore::BitmapImage::resetAnimation): Deleted.
(WebCore::BitmapImage::decode): Deleted.
(WebCore::BitmapImage::callDecodingCallbacks): Deleted.
(WebCore::BitmapImage::imageFrameAvailableAtIndex): Deleted.
(WebCore::BitmapImage::colorSpace): Deleted.
(WebCore::BitmapImage::decodeCountForTesting const): Deleted.
(WebCore::BitmapImage::lastDecodingOptions const): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp: Added.
(WebCore::BitmapImageSource::create):
(WebCore::BitmapImageSource::BitmapImageSource):
(WebCore::BitmapImageSource::decoder const):
(WebCore::BitmapImageSource::frameAnimator const):
(WebCore::BitmapImageSource::workQueue const):
(WebCore::BitmapImageSource::encodedDataStatusChanged):
(WebCore::BitmapImageSource::dataChanged):
(WebCore::BitmapImageSource::destroyDecodedData):
(WebCore::BitmapImageSource::decodedSizeChanged):
(WebCore::BitmapImageSource::decodedSizeIncreased):
(WebCore::BitmapImageSource::decodedSizeDecreased):
(WebCore::BitmapImageSource::decodedSizeReset):
(WebCore::BitmapImageSource::destroyNativeImageAtIndex):
(WebCore::BitmapImageSource::canDestroyDecodedData const):
(WebCore::BitmapImageSource::didDecodeProperties):
(WebCore::BitmapImageSource::clearFrameBufferCache):
(WebCore::BitmapImageSource::setData):
(WebCore::BitmapImageSource::resetData):
(WebCore::BitmapImageSource::startAnimation):
(WebCore::BitmapImageSource::stopAnimation):
(WebCore::BitmapImageSource::resetAnimation):
(WebCore::BitmapImageSource::isAnimated const):
(WebCore::BitmapImageSource::isAnimating const):
(WebCore::BitmapImageSource::isAnimationAllowed const):
(WebCore::BitmapImageSource::hasEverAnimated const):
(WebCore::BitmapImageSource::isLargeForDecoding const):
(WebCore::BitmapImageSource::isDecodingWorkQueueIdle const):
(WebCore::BitmapImageSource::stopDecodingWorkQueue):
(WebCore::BitmapImageSource::isPendingDecodingAtIndex const):
(WebCore::BitmapImageSource::isCompatibleWithOptionsAtIndex const):
(WebCore::BitmapImageSource::decode):
(WebCore::BitmapImageSource::callDecodeCallbacks):
(WebCore::BitmapImageSource::imageFrameAtIndexAvailable):
(WebCore::BitmapImageSource::imageFrameDecodeAtIndexHasFinished):
(WebCore::BitmapImageSource::currentFrameIndex const):
(WebCore::BitmapImageSource::cacheMetadataAtIndex):
(WebCore::BitmapImageSource::cacheNativeImageAtIndex):
(WebCore::BitmapImageSource::frameAtIndex const):
(WebCore::BitmapImageSource::frameAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::requestNativeImageAtIndex):
(WebCore::BitmapImageSource::requestNativeImageAtIndexIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexRequestIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexForDrawing):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
(WebCore::BitmapImageSource::nativeImageAtIndex):
(WebCore::BitmapImageSource::preTransformedNativeImageAtIndex):
(WebCore::BitmapImageSource::imageMetadata const):
(WebCore::BitmapImageSource::primaryNativeImageMetadata const):
(WebCore::BitmapImageSource::primaryImageFrameMetadata const):
(WebCore::BitmapImageSource::encodedDataStatus const):
(WebCore::BitmapImageSource::size const):
(WebCore::BitmapImageSource::sourceSize const):
(WebCore::BitmapImageSource::densityCorrectedSize const):
(WebCore::BitmapImageSource::orientation const):
(WebCore::BitmapImageSource::primaryFrameIndex const):
(WebCore::BitmapImageSource::frameCount const):
(WebCore::BitmapImageSource::repetitionCount const):
(WebCore::BitmapImageSource::colorSpace const):
(WebCore::BitmapImageSource::singlePixelSolidColor const):
(WebCore::BitmapImageSource::uti const):
(WebCore::BitmapImageSource::filenameExtension const):
(WebCore::BitmapImageSource::accessibilityDescription const):
(WebCore::BitmapImageSource::hotSpot const):
(WebCore::BitmapImageSource::maximumSubsamplingLevel const):
(WebCore::BitmapImageSource::subsamplingLevelForScaleFactor):
(WebCore::BitmapImageSource::frameSizeAtIndex const):
(WebCore::BitmapImageSource::frameDurationAtIndex const):
(WebCore::BitmapImageSource::frameOrientationAtIndex const):
(WebCore::BitmapImageSource::frameDecodingStatusAtIndex const):
(WebCore::BitmapImageSource::sourceUTF8 const):
(WebCore::BitmapImageSource::setMinimumDecodingDurationForTesting):
(WebCore::BitmapImageSource::dump const):
* Source/WebCore/platform/graphics/BitmapImageSource.h: Added.
(WebCore::BitmapImageSource::decodedSize const):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::decodeCountForTesting const):
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::drawPattern):
(WebCore::Image::drawTiled):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::frameCount const):
(WebCore::Image::nativeImageAtIndex):
(WebCore::Image::currentNativeImage):
(WebCore::Image::currentPreTransformedNativeImage):
(WebCore::Image::hasSolidColor):
(WebCore::Image::singlePixelSolidColor const):
(WebCore::Image::nativeImageForCurrentFrame): Deleted.
(WebCore::Image::preTransformedNativeImageForCurrentFrame): Deleted.
(WebCore::Image::nativeImageAtIndexCacheIfNeeded): Deleted.
(WebCore::Image::notSolidColor): Deleted.
* Source/WebCore/platform/graphics/ImageAdapter.cpp:
(WebCore::ImageAdapter::nativeImageOfSize):
(WebCore::ImageAdapter::allNativeImages):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::frameOrientationAtIndex const):
(WebCore::ImageDecoder::densityCorrectedSizeAtIndex const):
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::ImageFrame):
(WebCore::ImageFrame::size const): Deleted.
(WebCore::ImageFrame::singlePixelSolidColor const): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::setSize):
(WebCore::ImageFrame::size const):
(WebCore::ImageFrame::decodingOptions const):
* Source/WebCore/platform/graphics/ImageFrameAnimator.cpp: Added.
(WebCore::ImageFrameAnimator::create):
(WebCore::ImageFrameAnimator::ImageFrameAnimator):
(WebCore::ImageFrameAnimator::~ImageFrameAnimator):
(WebCore::ImageFrameAnimator::destroyDecodedData):
(WebCore::ImageFrameAnimator::startTimer):
(WebCore::ImageFrameAnimator::clearTimer):
(WebCore::ImageFrameAnimator::timerFired):
(WebCore::ImageFrameAnimator::imageFrameDecodeAtIndexHasFinished):
(WebCore::ImageFrameAnimator::startAnimation):
(WebCore::ImageFrameAnimator::advanceAnimation):
(WebCore::ImageFrameAnimator::stopAnimation):
(WebCore::ImageFrameAnimator::resetAnimation):
(WebCore::ImageFrameAnimator::isAnimationAllowed const):
(WebCore::ImageFrameAnimator::sourceUTF8 const):
(WebCore::ImageFrameAnimator::dump const):
* Source/WebCore/platform/graphics/ImageFrameAnimator.h: Added.
(WebCore::ImageFrameAnimator::isAnimating const):
(WebCore::ImageFrameAnimator::hasEverAnimated const):
(WebCore::ImageFrameAnimator::currentFrameIndex const):
(WebCore::ImageFrameAnimator::nextFrameIndex const):
* Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp: Added.
(WebCore::ImageFrameWorkQueue::create):
(WebCore::ImageFrameWorkQueue::ImageFrameWorkQueue):
(WebCore::ImageFrameWorkQueue::requestQueue):
(WebCore::ImageFrameWorkQueue::start):
(WebCore::ImageFrameWorkQueue::dispatch):
(WebCore::ImageFrameWorkQueue::stop):
(WebCore::ImageFrameWorkQueue::isPendingDecodingAtIndex const):
(WebCore::ImageFrameWorkQueue::dump const):
* Source/WebCore/platform/graphics/ImageFrameWorkQueue.h: Added.
(WebCore::ImageFrameWorkQueue::isIdle const):
(WebCore::ImageFrameWorkQueue::setMinimumDecodingDurationForTesting):
(WebCore::ImageFrameWorkQueue::protectedSource const):
(WebCore::ImageFrameWorkQueue::decodeQueue):
(WebCore::ImageFrameWorkQueue::minimumDecodingDurationForTesting const):
* Source/WebCore/platform/graphics/ImageObserver.h:
(WebCore::ImageObserver::canDestroyDecodedData const):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::primaryNativeImageForDrawing):
(WebCore::ImageSource::currentNativeImageForDrawing):
(WebCore::ImageSource::hasSolidColor const):
(WebCore::ImageSource::ImageSource): Deleted.
(WebCore::ImageSource::~ImageSource): Deleted.
(WebCore::ImageSource::ensureDecoderAvailable): Deleted.
(WebCore::ImageSource::setData): Deleted.
(WebCore::ImageSource::resetData): Deleted.
(WebCore::ImageSource::dataChanged): Deleted.
(WebCore::ImageSource::isAllDataReceived): Deleted.
(WebCore::ImageSource::destroyDecodedData): Deleted.
(WebCore::ImageSource::destroyIncompleteDecodedData): Deleted.
(WebCore::ImageSource::clearFrameBufferCache): Deleted.
(WebCore::ImageSource::encodedDataStatusChanged): Deleted.
(WebCore::ImageSource::decodedSizeChanged): Deleted.
(WebCore::ImageSource::decodedSizeIncreased): Deleted.
(WebCore::ImageSource::decodedSizeDecreased): Deleted.
(WebCore::ImageSource::decodedSizeReset): Deleted.
(WebCore::ImageSource::didDecodeProperties): Deleted.
(WebCore::ImageSource::growFrames): Deleted.
(WebCore::ImageSource::setNativeImage): Deleted.
(WebCore::ImageSource::cacheMetadataAtIndex): Deleted.
(WebCore::ImageSource::cachePlatformImageAtIndex): Deleted.
(WebCore::ImageSource::cachePlatformImageAtIndexAsync): Deleted.
(WebCore::ImageSource::decodingQueue): Deleted.
(WebCore::ImageSource::frameRequestQueue): Deleted.
(WebCore::ImageSource::canUseAsyncDecoding): Deleted.
(WebCore::ImageSource::startAsyncDecodingQueue): Deleted.
(WebCore::ImageSource::requestFrameAsyncDecodingAtIndex): Deleted.
(WebCore::ImageSource::isAsyncDecodingQueueIdle const): Deleted.
(WebCore::ImageSource::stopAsyncDecodingQueue): Deleted.
(WebCore::ImageSource::frameAtIndexCacheIfNeeded): Deleted.
(WebCore::ImageSource::clearMetadata): Deleted.
(WebCore::ImageSource::sourceURL const): Deleted.
(WebCore::ImageSource::mimeType const): Deleted.
(WebCore::ImageSource::expectedContentLength const): Deleted.
(WebCore::ImageSource::metadataCacheIfNeeded): Deleted.
(WebCore::ImageSource::firstFrameMetadataCacheIfNeeded): Deleted.
(WebCore::ImageSource::encodedDataStatus): Deleted.
(WebCore::ImageSource::frameCount): Deleted.
(WebCore::ImageSource::primaryFrameIndex): Deleted.
(WebCore::ImageSource::repetitionCount): Deleted.
(WebCore::ImageSource::uti): Deleted.
(WebCore::ImageSource::filenameExtension): Deleted.
(WebCore::ImageSource::accessibilityDescription): Deleted.
(WebCore::ImageSource::hotSpot): Deleted.
(WebCore::ImageSource::orientation): Deleted.
(WebCore::ImageSource::densityCorrectedSize): Deleted.
(WebCore::ImageSource::size): Deleted.
(WebCore::ImageSource::sourceSize): Deleted.
(WebCore::ImageSource::singlePixelSolidColor): Deleted.
(WebCore::ImageSource::maximumSubsamplingLevel): Deleted.
(WebCore::ImageSource::frameIsBeingDecodedAndIsCompatibleWithOptionsAtIndex): Deleted.
(WebCore::ImageSource::frameDecodingStatusAtIndex): Deleted.
(WebCore::ImageSource::frameHasAlphaAtIndex): Deleted.
(WebCore::ImageSource::frameHasFullSizeNativeImageAtIndex): Deleted.
(WebCore::ImageSource::frameHasDecodedNativeImageCompatibleWithOptionsAtIndex): Deleted.
(WebCore::ImageSource::frameSubsamplingLevelAtIndex): Deleted.
(WebCore::ImageSource::frameSizeAtIndex): Deleted.
(WebCore::ImageSource::frameBytesAtIndex): Deleted.
(WebCore::ImageSource::frameDurationAtIndex): Deleted.
(WebCore::ImageSource::frameOrientationAtIndex): Deleted.
(WebCore::ImageSource::createFrameImageAtIndex): Deleted.
(WebCore::ImageSource::frameImageAtIndex): Deleted.
(WebCore::ImageSource::frameImageAtIndexCacheIfNeeded): Deleted.
(WebCore::ImageSource::dump): Deleted.
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::dataChanged):
(WebCore::ImageSource::destroyDecodedData):
(WebCore::ImageSource::startAnimation):
(WebCore::ImageSource::stopAnimation):
(WebCore::ImageSource::resetAnimation):
(WebCore::ImageSource::isAnimated const):
(WebCore::ImageSource::isAnimating const):
(WebCore::ImageSource::hasEverAnimated const):
(WebCore::ImageSource::isLargeForDecoding const):
(WebCore::ImageSource::stopDecodingWorkQueue):
(WebCore::ImageSource::decode):
(WebCore::ImageSource::currentFrameIndex const):
(WebCore::ImageSource::currentImageFrame):
(WebCore::ImageSource::currentNativeImage):
(WebCore::ImageSource::currentPreTransformedNativeImage):
(WebCore::ImageSource::nativeImageAtIndex):
(WebCore::ImageSource::sourceSize const):
(WebCore::ImageSource::hasDensityCorrectedSize const):
(WebCore::ImageSource::orientation const):
(WebCore::ImageSource::primaryFrameIndex const):
(WebCore::ImageSource::frameCount const):
(WebCore::ImageSource::uti const):
(WebCore::ImageSource::filenameExtension const):
(WebCore::ImageSource::accessibilityDescription const):
(WebCore::ImageSource::hotSpot const):
(WebCore::ImageSource::subsamplingLevelForScaleFactor):
(WebCore::ImageSource::frameDurationAtIndex const):
(WebCore::ImageSource::frameOrientationAtIndex const):
(WebCore::ImageSource::frameDecodingStatusAtIndex const):
(WebCore::ImageSource::decodeCountForTesting const):
(WebCore::ImageSource::setMinimumDecodingDurationForTesting):
(WebCore::ImageSource::setClearDecoderAfterAsyncFrameRequestForTesting):
(WebCore::ImageSource::setAsyncDecodingEnabledForTesting):
(WebCore::ImageSource::isAsyncDecodingEnabledForTesting const):
(WebCore::ImageSource::dump const):
(): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/NativeImageSource.cpp: Copied from Source/WebCore/platform/graphics/ImageAdapter.cpp.
(WebCore::NativeImageSource::create):
(WebCore::NativeImageSource::NativeImageSource):
(WebCore::NativeImageSource::dump const):
* Source/WebCore/platform/graphics/NativeImageSource.h: Copied from Source/WebCore/platform/graphics/ImageAdapter.cpp.
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
(WebCore::ImageDecoderAVFObjC::frameMetadataAtIndex const): Deleted.
(WebCore::ImageDecoderAVFObjC::frameAllowSubsamplingAtIndex const): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToImage):
* Source/WebCore/platform/graphics/cairo/GraphicsContextGLCairo.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cairo/NativeImageCairo.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::frameOrientationAtIndex const):
(WebCore::ImageDecoderCG::densityCorrectedSizeAtIndex const):
(WebCore::ImageDecoderCG::frameMetadataAtIndex const): Deleted.
(WebCore::ImageDecoderCG::frameAllowSubsamplingAtIndex const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
(WebCore::ImageDecoderGStreamer::uti const): Deleted.
(WebCore::ImageDecoderGStreamer::frameMetadataAtIndex const): Deleted.
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.h:
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/graphics/mac/ImageAdapterMac.mm:
(WebCore::ImageAdapter::snapshotNSImage):
* Source/WebCore/platform/graphics/skia/GraphicsContextGLSkia.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.cpp:
(WebCore::GraphicsLayerTextureMapper::setContentsToImage):
* Source/WebCore/platform/graphics/texmap/TextureMapperTile.cpp:
(WebCore::TextureMapperTile::updateContents):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::setContentsToImage):
* Source/WebCore/platform/gtk/DragImageGtk.cpp:
(WebCore::createDragImageFromImage):
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.h:
* Source/WebCore/platform/win/cairo/DragImageWinCairo.cpp:
(WebCore::createDragImageFromImage):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::isDirectlyCompositedImage const):
(WebCore::RenderLayerBacking::isUnscaledBitmapOnly const):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::createFilterEffect const):
* Source/WebCore/svg/graphics/SVGImageForContainer.cpp:
(WebCore::SVGImageForContainer::currentNativeImage):
(WebCore::SVGImageForContainer::nativeImageForCurrentFrame): Deleted.
* Source/WebCore/svg/graphics/SVGImageForContainer.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setImageFrameDecodingDuration):
(WebCore::Internals::imageLastDecodingOptions):
* Source/WebKit/UIProcess/API/glib/IconDatabase.cpp:
(WebKit::IconDatabase::loadIconForPageURL):
(WebKit::IconDatabase::setIconForPageURL):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::setCursor):
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVF.cpp:
(WebKit::RemoteImageDecoderAVF::frameMetadataAtIndex const): Deleted.
(WebKit::RemoteImageDecoderAVF::frameAllowSubsamplingAtIndex const): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVF.h:
* Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:

Canonical link: <a href="https://commits.webkit.org/276827@main">https://commits.webkit.org/276827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9f26206e0f89238ec63f44dc0db205209d9ddc2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45741 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24868 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48411 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41776 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48047 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22266 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37453 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46319 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21946 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39461 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18608 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40551 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3785 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42050 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40891 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50165 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20732 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17246 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44567 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39612 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43420 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10170 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22396 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/NativeImageSource.cpp:
```diff
@@ -0,0 +1,48 @@
+
+NativeImageSource::NativeImageSource(Ref<NativeImage>&& nativeImage)
+{
+    m_frame.m_size = nativeImage->size();
```
Rather than requiring `friend`, why not add a constructor to ImageFrame that takes a NativeImage and does these assignments?

shallawa replied:
Fixed.

weinig commented on Source/WebCore/platform/graphics/NativeImage.h:
```diff
@@ -51,7 +51,7 @@ class NativeImage final : public RenderingResource {
 
```
I think that keeping this called NativeImage, but not having it inherit from Image will probably continue to be confusing. Perhaps leaning into its "resource"-ness and going with `NativeImageResource`? Anything to avoid the `Image` suffix.

shallawa replied:
Yes I agree. I will address this in a separate patch.

shallawa committed:
Decouple BitmapImage sources and main functions into separate classes
https://bugs.webkit.org/show_bug.cgi?id=265363
rdar://118819338

Reviewed by Cameron McCormack.

The BitmapImage and ImageSource manage the operations and the queries of the
bitmap image. The current interface and the code structure of these two classes
and the relationship between them make it difficult and risky to hack.

This purpose of this refactor is to provide a clearer interface and to build a
a better class hierarchy and association. This will allow distributing the
operations among better encapsulated classes.

Here is the new class hierarchy and association:

  Image: The super class for all kinds of images in WebKit.

  ImageAdapter: It converts an Image to a platform specific format.

  ImageDecoder: A super class for the system and WebKit image decoders

  ImageSource: A super class BitmapImage source. It can be from a CachedImage or
  from a NativeImage.

  BitmapImageSource: Receives the encoded data of an image and provides all the
  frames decoded NativeImage and metadata. It holds and coordinates the interactions
  amog BitmapImage, ImageFrameAnimator and ImageFrameWorkQueue.

  NativeImageSource: It wraps an already created NativeImage in a BitmapImage.

  ImageFrameAnimator: It advances the current frame index of an animated image when.
  its timer fires. The timer is set with the duration of the frame. It requests
  decoding the NativeImage asynchronously if this is allowed.

  ImageFrameWorkQueue: It decodes the NativeImage of an ImageFrame asynchronously
  on a separate WorkQueue. Once the NativeImage is ready, it asks BitmapImageSource
  to cache it on the main thread. BitmapImageSource will notifies the observer
  that a new frame is ready for rendering.

  BitmapImage: This is a sub-class of Image which mainly handles the layout and
  the drawing of a bitmap image to a GraphicsContext.

An important change of this PR is the DecodingMode and AllowImageSubsampling are
calculated in advance before calling BitmapImage::draw(). So Settings are no
longer passed or updated by BitmapImage before calling its draw() method.

* LayoutTests/fast/images/async-image-background-image-repeated.html:
Make the frame decoding duration time longer than the animation duration time so
the decoding always finishes after the animation timer fires.

* LayoutTests/fast/images/async-image-composited-show.html:
Make this test do the decoding asynchronously. The debug settings were missing.

* LayoutTests/fast/images/async-image-intersect-different-size-for-drawing-expected.txt:
If the decodingMode is Synchronous, no sizeForDrawing should be set.

* LayoutTests/fast/images/decode-render-animated-image-expected.html:
* LayoutTests/fast/images/decode-render-animated-image.html:
Make the test relies solely on the `decode()` method instead of the mix between
the debug event "webkitImageFrameReady" and the `decode()` method.

* LayoutTests/platform/ios-wk2/fast/images/async-image-intersect-different-size-for-drawing-expected.txt:
If the decodingMode is Synchronous, no sizeForDrawing should be set.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::decode):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::removeAllClientsWaitingForAsyncDecoding):
(WebCore::CachedImage::CachedImageObserver::canDestroyDecodedData const):
(WebCore::CachedImage::canDestroyDecodedData const):
(WebCore::CachedImage::CachedImageObserver::canDestroyDecodedData): Deleted.
(WebCore::CachedImage::canDestroyDecodedData): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::canDestroyDecodedData const):
(WebCore::CachedImageClient::canDestroyDecodedData): Deleted.
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::createDragImageFromImage):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::create):
(WebCore::BitmapImage::BitmapImage):
(WebCore::BitmapImage::dataChanged):
(WebCore::BitmapImage::destroyDecodedData):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::drawPattern):
(WebCore::BitmapImage::dump const):
(WebCore::BitmapImage::~BitmapImage): Deleted.
(WebCore::BitmapImage::destroyDecodedDataIfNecessary): Deleted.
(WebCore::BitmapImage::setCurrentFrameDecodingStatusIfNecessary): Deleted.
(WebCore::BitmapImage::nativeImageAtIndexCacheIfNeeded): Deleted.
(WebCore::BitmapImage::nativeImage): Deleted.
(WebCore::BitmapImage::nativeImageForCurrentFrame): Deleted.
(WebCore::BitmapImage::preTransformedNativeImageForCurrentFrame): Deleted.
(WebCore::BitmapImage::notSolidColor): Deleted.
(WebCore::drawNativeImage): Deleted.
(WebCore::BitmapImage::shouldAnimate const): Deleted.
(WebCore::BitmapImage::canAnimate const): Deleted.
(WebCore::BitmapImage::canUseAsyncDecodingForLargeImages const): Deleted.
(WebCore::BitmapImage::shouldUseAsyncDecodingForAnimatedImages const): Deleted.
(WebCore::BitmapImage::clearTimer): Deleted.
(WebCore::BitmapImage::startTimer): Deleted.
(WebCore::BitmapImage::subsamplingLevelForScaleFactor): Deleted.
(WebCore::BitmapImage::canDestroyDecodedData): Deleted.
(WebCore::BitmapImage::internalStartAnimation): Deleted.
(WebCore::BitmapImage::advanceAnimation): Deleted.
(WebCore::BitmapImage::internalAdvanceAnimation): Deleted.
(WebCore::BitmapImage::isAnimating const): Deleted.
(WebCore::BitmapImage::stopAnimation): Deleted.
(WebCore::BitmapImage::resetAnimation): Deleted.
(WebCore::BitmapImage::decode): Deleted.
(WebCore::BitmapImage::callDecodingCallbacks): Deleted.
(WebCore::BitmapImage::imageFrameAvailableAtIndex): Deleted.
(WebCore::BitmapImage::colorSpace): Deleted.
(WebCore::BitmapImage::decodeCountForTesting const): Deleted.
(WebCore::BitmapImage::lastDecodingOptions const): Deleted.
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp: Added.
(WebCore::BitmapImageSource::create):
(WebCore::BitmapImageSource::BitmapImageSource):
(WebCore::BitmapImageSource::decoder const):
(WebCore::BitmapImageSource::frameAnimator const):
(WebCore::BitmapImageSource::workQueue const):
(WebCore::BitmapImageSource::encodedDataStatusChanged):
(WebCore::BitmapImageSource::dataChanged):
(WebCore::BitmapImageSource::destroyDecodedData):
(WebCore::BitmapImageSource::decodedSizeChanged):
(WebCore::BitmapImageSource::decodedSizeIncreased):
(WebCore::BitmapImageSource::decodedSizeDecreased):
(WebCore::BitmapImageSource::decodedSizeReset):
(WebCore::BitmapImageSource::destroyNativeImageAtIndex):
(WebCore::BitmapImageSource::canDestroyDecodedData const):
(WebCore::BitmapImageSource::didDecodeProperties):
(WebCore::BitmapImageSource::clearFrameBufferCache):
(WebCore::BitmapImageSource::setData):
(WebCore::BitmapImageSource::resetData):
(WebCore::BitmapImageSource::startAnimation):
(WebCore::BitmapImageSource::stopAnimation):
(WebCore::BitmapImageSource::resetAnimation):
(WebCore::BitmapImageSource::isAnimated const):
(WebCore::BitmapImageSource::isAnimating const):
(WebCore::BitmapImageSource::isAnimationAllowed const):
(WebCore::BitmapImageSource::hasEverAnimated const):
(WebCore::BitmapImageSource::isLargeForDecoding const):
(WebCore::BitmapImageSource::isDecodingWorkQueueIdle const):
(WebCore::BitmapImageSource::stopDecodingWorkQueue):
(WebCore::BitmapImageSource::isPendingDecodingAtIndex const):
(WebCore::BitmapImageSource::isCompatibleWithOptionsAtIndex const):
(WebCore::BitmapImageSource::decode):
(WebCore::BitmapImageSource::callDecodeCallbacks):
(WebCore::BitmapImageSource::imageFrameAtIndexAvailable):
(WebCore::BitmapImageSource::imageFrameDecodeAtIndexHasFinished):
(WebCore::BitmapImageSource::currentFrameIndex const):
(WebCore::BitmapImageSource::cacheMetadataAtIndex):
(WebCore::BitmapImageSource::cacheNativeImageAtIndex):
(WebCore::BitmapImageSource::frameAtIndex const):
(WebCore::BitmapImageSource::frameAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::requestNativeImageAtIndex):
(WebCore::BitmapImageSource::requestNativeImageAtIndexIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexRequestIfNeeded):
(WebCore::BitmapImageSource::nativeImageAtIndexForDrawing):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
(WebCore::BitmapImageSource::nativeImageAtIndex):
(WebCore::BitmapImageSource::preTransformedNativeImageAtIndex):
(WebCore::BitmapImageSource::imageMetadata const):
(WebCore::BitmapImageSource::primaryNativeImageMetadata const):
(WebCore::BitmapImageSource::primaryImageFrameMetadata const):
(WebCore::BitmapImageSource::encodedDataStatus const):
(WebCore::BitmapImageSource::size const):
(WebCore::BitmapImageSource::sourceSize const):
(WebCore::BitmapImageSource::densityCorrectedSize const):
(WebCore::BitmapImageSource::orientation const):
(WebCore::BitmapImageSource::primaryFrameIndex const):
(WebCore::BitmapImageSource::frameCount const):
(WebCore::BitmapImageSource::repetitionCount const):
(WebCore::BitmapImageSource::colorSpace const):
(WebCore::BitmapImageSource::singlePixelSolidColor const):
(WebCore::BitmapImageSource::uti const):
(WebCore::BitmapImageSource::filenameExtension const):
(WebCore::BitmapImageSource::accessibilityDescription const):
(WebCore::BitmapImageSource::hotSpot const):
(WebCore::BitmapImageSource::maximumSubsamplingLevel const):
(WebCore::BitmapImageSource::subsamplingLevelForScaleFactor):
(WebCore::BitmapImageSource::frameSizeAtIndex const):
(WebCore::BitmapImageSource::frameDurationAtIndex const):
(WebCore::BitmapImageSource::frameOrientationAtIndex const):
(WebCore::BitmapImageSource::frameDecodingStatusAtIndex const):
(WebCore::BitmapImageSource::sourceUTF8 const):
(WebCore::BitmapImageSource::setMinimumDecodingDurationForTesting):
(WebCore::BitmapImageSource::dump const):
* Source/WebCore/platform/graphics/BitmapImageSource.h: Added.
(WebCore::BitmapImageSource::decodedSize const):
(WebCore::BitmapImageSource::nativeImageAtIndexCacheIfNeeded):
(WebCore::BitmapImageSource::decodeCountForTesting const):
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::drawPattern):
(WebCore::Image::drawTiled):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::frameCount const):
(WebCore::Image::nativeImageAtIndex):
(WebCore::Image::currentNativeImage):
(WebCore::Image::currentPreTransformedNativeImage):
(WebCore::Image::hasSolidColor):
(WebCore::Image::singlePixelSolidColor const):
(WebCore::Image::nativeImageForCurrentFrame): Deleted.
(WebCore::Image::preTransformedNativeImageForCurrentFrame): Deleted.
(WebCore::Image::nativeImageAtIndexCacheIfNeeded): Deleted.
(WebCore::Image::notSolidColor): Deleted.
* Source/WebCore/platform/graphics/ImageAdapter.cpp:
(WebCore::ImageAdapter::nativeImageOfSize):
(WebCore::ImageAdapter::allNativeImages):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::frameOrientationAtIndex const):
(WebCore::ImageDecoder::densityCorrectedSizeAtIndex const):
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::ImageFrame):
(WebCore::ImageFrame::size const): Deleted.
(WebCore::ImageFrame::singlePixelSolidColor const): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::setSize):
(WebCore::ImageFrame::size const):
(WebCore::ImageFrame::decodingOptions const):
* Source/WebCore/platform/graphics/ImageFrameAnimator.cpp: Added.
(WebCore::ImageFrameAnimator::create):
(WebCore::ImageFrameAnimator::ImageFrameAnimator):
(WebCore::ImageFrameAnimator::~ImageFrameAnimator):
(WebCore::ImageFrameAnimator::destroyDecodedData):
(WebCore::ImageFrameAnimator::startTimer):
(WebCore::ImageFrameAnimator::clearTimer):
(WebCore::ImageFrameAnimator::timerFired):
(WebCore::ImageFrameAnimator::imageFrameDecodeAtIndexHasFinished):
(WebCore::ImageFrameAnimator::startAnimation):
(WebCore::ImageFrameAnimator::advanceAnimation):
(WebCore::ImageFrameAnimator::stopAnimation):
(WebCore::ImageFrameAnimator::resetAnimation):
(WebCore::ImageFrameAnimator::isAnimationAllowed const):
(WebCore::ImageFrameAnimator::sourceUTF8 const):
(WebCore::ImageFrameAnimator::dump const):
* Source/WebCore/platform/graphics/ImageFrameAnimator.h: Added.
(WebCore::ImageFrameAnimator::isAnimating const):
(WebCore::ImageFrameAnimator::hasEverAnimated const):
(WebCore::ImageFrameAnimator::currentFrameIndex const):
(WebCore::ImageFrameAnimator::nextFrameIndex const):
* Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp: Added.
(WebCore::ImageFrameWorkQueue::create):
(WebCore::ImageFrameWorkQueue::ImageFrameWorkQueue):
(WebCore::ImageFrameWorkQueue::requestQueue):
(WebCore::ImageFrameWorkQueue::start):
(WebCore::ImageFrameWorkQueue::dispatch):
(WebCore::ImageFrameWorkQueue::stop):
(WebCore::ImageFrameWorkQueue::isPendingDecodingAtIndex const):
(WebCore::ImageFrameWorkQueue::dump const):
* Source/WebCore/platform/graphics/ImageFrameWorkQueue.h: Added.
(WebCore::ImageFrameWorkQueue::isIdle const):
(WebCore::ImageFrameWorkQueue::setMinimumDecodingDurationForTesting):
(WebCore::ImageFrameWorkQueue::protectedSource const):
(WebCore::ImageFrameWorkQueue::decodeQueue):
(WebCore::ImageFrameWorkQueue::minimumDecodingDurationForTesting const):
* Source/WebCore/platform/graphics/ImageObserver.h:
(WebCore::ImageObserver::canDestroyDecodedData const):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::primaryNativeImageForDrawing):
(WebCore::ImageSource::currentNativeImageForDrawing):
(WebCore::ImageSource::hasSolidColor const):
(WebCore::ImageSource::ImageSource): Deleted.
(WebCore::ImageSource::~ImageSource): Deleted.
(WebCore::ImageSource::ensureDecoderAvailable): Deleted.
(WebCore::ImageSource::setData): Deleted.
(WebCore::ImageSource::resetData): Deleted.
(WebCore::ImageSource::dataChanged): Deleted.
(WebCore::ImageSource::isAllDataReceived): Deleted.
(WebCore::ImageSource::destroyDecodedData): Deleted.
(WebCore::ImageSource::destroyIncompleteDecodedData): Deleted.
(WebCore::ImageSource::clearFrameBufferCache): Deleted.
(WebCore::ImageSource::encodedDataStatusChanged): Deleted.
(WebCore::ImageSource::decodedSizeChanged): Deleted.
(WebCore::ImageSource::decodedSizeIncreased): Deleted.
(WebCore::ImageSource::decodedSizeDecreased): Deleted.
(WebCore::ImageSource::decodedSizeReset): Deleted.
(WebCore::ImageSource::didDecodeProperties): Deleted.
(WebCore::ImageSource::growFrames): Deleted.
(WebCore::ImageSource::setNativeImage): Deleted.
(WebCore::ImageSource::cacheMetadataAtIndex): Deleted.
(WebCore::ImageSource::cachePlatformImageAtIndex): Deleted.
(WebCore::ImageSource::cachePlatformImageAtIndexAsync): Deleted.
(WebCore::ImageSource::decodingQueue): Deleted.
(WebCore::ImageSource::frameRequestQueue): Deleted.
(WebCore::ImageSource::canUseAsyncDecoding): Deleted.
(WebCore::ImageSource::startAsyncDecodingQueue): Deleted.
(WebCore::ImageSource::requestFrameAsyncDecodingAtIndex): Deleted.
(WebCore::ImageSource::isAsyncDecodingQueueIdle const): Deleted.
(WebCore::ImageSource::stopAsyncDecodingQueue): Deleted.
(WebCore::ImageSource::frameAtIndexCacheIfNeeded): Deleted.
(WebCore::ImageSource::clearMetadata): Deleted.
(WebCore::ImageSource::sourceURL const): Deleted.
(WebCore::ImageSource::mimeType const): Deleted.
(WebCore::ImageSource::expectedContentLength const): Deleted.
(WebCore::ImageSource::metadataCacheIfNeeded): Deleted.
(WebCore::ImageSource::firstFrameMetadataCacheIfNeeded): Deleted.
(WebCore::ImageSource::encodedDataStatus): Deleted.
(WebCore::ImageSource::frameCount): Deleted.
(WebCore::ImageSource::primaryFrameIndex): Deleted.
(WebCore::ImageSource::repetitionCount): Deleted.
(WebCore::ImageSource::uti): Deleted.
(WebCore::ImageSource::filenameExtension): Deleted.
(WebCore::ImageSource::accessibilityDescription): Deleted.
(WebCore::ImageSource::hotSpot): Deleted.
(WebCore::ImageSource::orientation): Deleted.
(WebCore::ImageSource::densityCorrectedSize): Deleted.
(WebCore::ImageSource::size): Deleted.
(WebCore::ImageSource::sourceSize): Deleted.
(WebCore::ImageSource::singlePixelSolidColor): Deleted.
(WebCore::ImageSource::maximumSubsamplingLevel): Deleted.
(WebCore::ImageSource::frameIsBeingDecodedAndIsCompatibleWithOptionsAtIndex): Deleted.
(WebCore::ImageSource::frameDecodingStatusAtIndex): Deleted.
(WebCore::ImageSource::frameHasAlphaAtIndex): Deleted.
(WebCore::ImageSource::frameHasFullSizeNativeImageAtIndex): Deleted.
(WebCore::ImageSource::frameHasDecodedNativeImageCompatibleWithOptionsAtIndex): Deleted.
(WebCore::ImageSource::frameSubsamplingLevelAtIndex): Deleted.
(WebCore::ImageSource::frameSizeAtIndex): Deleted.
(WebCore::ImageSource::frameBytesAtIndex): Deleted.
(WebCore::ImageSource::frameDurationAtIndex): Deleted.
(WebCore::ImageSource::frameOrientationAtIndex): Deleted.
(WebCore::ImageSource::createFrameImageAtIndex): Deleted.
(WebCore::ImageSource::frameImageAtIndex): Deleted.
(WebCore::ImageSource::frameImageAtIndexCacheIfNeeded): Deleted.
(WebCore::ImageSource::dump): Deleted.
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::dataChanged):
(WebCore::ImageSource::destroyDecodedData):
(WebCore::ImageSource::startAnimation):
(WebCore::ImageSource::stopAnimation):
(WebCore::ImageSource::resetAnimation):
(WebCore::ImageSource::isAnimated const):
(WebCore::ImageSource::isAnimating const):
(WebCore::ImageSource::hasEverAnimated const):
(WebCore::ImageSource::isLargeForDecoding const):
(WebCore::ImageSource::stopDecodingWorkQueue):
(WebCore::ImageSource::decode):
(WebCore::ImageSource::currentFrameIndex const):
(WebCore::ImageSource::currentImageFrame):
(WebCore::ImageSource::currentNativeImage):
(WebCore::ImageSource::currentPreTransformedNativeImage):
(WebCore::ImageSource::nativeImageAtIndex):
(WebCore::ImageSource::sourceSize const):
(WebCore::ImageSource::hasDensityCorrectedSize const):
(WebCore::ImageSource::orientation const):
(WebCore::ImageSource::primaryFrameIndex const):
(WebCore::ImageSource::frameCount const):
(WebCore::ImageSource::uti const):
(WebCore::ImageSource::filenameExtension const):
(WebCore::ImageSource::accessibilityDescription const):
(WebCore::ImageSource::hotSpot const):
(WebCore::ImageSource::subsamplingLevelForScaleFactor):
(WebCore::ImageSource::frameDurationAtIndex const):
(WebCore::ImageSource::frameOrientationAtIndex const):
(WebCore::ImageSource::frameDecodingStatusAtIndex const):
(WebCore::ImageSource::decodeCountForTesting const):
(WebCore::ImageSource::setMinimumDecodingDurationForTesting):
(WebCore::ImageSource::setClearDecoderAfterAsyncFrameRequestForTesting):
(WebCore::ImageSource::setAsyncDecodingEnabledForTesting):
(WebCore::ImageSource::isAsyncDecodingEnabledForTesting const):
(WebCore::ImageSource::dump const):
(): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/NativeImageSource.cpp: Copied from Source/WebCore/platform/graphics/ImageAdapter.cpp.
(WebCore::NativeImageSource::create):
(WebCore::NativeImageSource::NativeImageSource):
(WebCore::NativeImageSource::dump const):
* Source/WebCore/platform/graphics/NativeImageSource.h: Copied from Source/WebCore/platform/graphics/ImageAdapter.cpp.
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/ImageDecoderAVFObjC.mm:
(WebCore::ImageDecoderAVFObjC::frameMetadataAtIndex const): Deleted.
(WebCore::ImageDecoderAVFObjC::frameAllowSubsamplingAtIndex const): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToImage):
* Source/WebCore/platform/graphics/cairo/GraphicsContextGLCairo.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cairo/NativeImageCairo.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/cg/GraphicsContextGLCG.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::frameOrientationAtIndex const):
(WebCore::ImageDecoderCG::densityCorrectedSizeAtIndex const):
(WebCore::ImageDecoderCG::frameMetadataAtIndex const): Deleted.
(WebCore::ImageDecoderCG::frameAllowSubsamplingAtIndex const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp:
(WebCore::ImageDecoderGStreamer::uti const): Deleted.
(WebCore::ImageDecoderGStreamer::frameMetadataAtIndex const): Deleted.
* Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.h:
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::ImageAdapter::gdkPixbuf):
(WebCore::ImageAdapter::gdkTexture):
* Source/WebCore/platform/graphics/mac/ImageAdapterMac.mm:
(WebCore::ImageAdapter::snapshotNSImage):
* Source/WebCore/platform/graphics/skia/GraphicsContextGLSkia.cpp:
(WebCore::GraphicsContextGLImageExtractor::extractImage):
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::NativeImage::singlePixelSolidColor const):
* Source/WebCore/platform/graphics/texmap/GraphicsLayerTextureMapper.cpp:
(WebCore::GraphicsLayerTextureMapper::setContentsToImage):
* Source/WebCore/platform/graphics/texmap/TextureMapperTile.cpp:
(WebCore::TextureMapperTile::updateContents):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::setContentsToImage):
* Source/WebCore/platform/gtk/DragImageGtk.cpp:
(WebCore::createDragImageFromImage):
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.h:
* Source/WebCore/platform/win/cairo/DragImageWinCairo.cpp:
(WebCore::createDragImageFromImage):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::isDirectlyCompositedImage const):
(WebCore::RenderLayerBacking::isUnscaledBitmapOnly const):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::createFilterEffect const):
* Source/WebCore/svg/graphics/SVGImageForContainer.cpp:
(WebCore::SVGImageForContainer::currentNativeImage):
(WebCore::SVGImageForContainer::nativeImageForCurrentFrame): Deleted.
* Source/WebCore/svg/graphics/SVGImageForContainer.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setImageFrameDecodingDuration):
(WebCore::Internals::imageLastDecodingOptions):
* Source/WebKit/UIProcess/API/glib/IconDatabase.cpp:
(WebKit::IconDatabase::loadIconForPageURL):
(WebKit::IconDatabase::setIconForPageURL):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::setCursor):
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVF.cpp:
(WebKit::RemoteImageDecoderAVF::frameMetadataAtIndex const): Deleted.
(WebKit::RemoteImageDecoderAVF::frameAllowSubsamplingAtIndex const): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVF.h:
* Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:

Canonical link: https://commits.webkit.org/276827@main
